### ISSUE TITLE
Binding life-cycle callbacks

### ISSUE THREAD

**Issue by mbest** [knockout/knockout#2165]
In the original spec for components (#1290), Steve described the possibility of adding lifecycle callbacks to components, but didn't want to "make the wrong guesses here." I think it's been enough time that we should figure out how to include such callbacks.

As part of the discussion on an "afterRender" callback, @chrisknoll [suggested](https://github.com/knockout/knockout/issues/1533#issuecomment-169897476) we might be better off making this lifecycle tracking generic to all bindings.

**Comment by mbest** [comment]
For comparison, here are lifecycle callbacks from other libraries, etc.:

- http://durandaljs.com/documentation/Hooking-Lifecycle-Callbacks
- https://angular.io/docs/ts/latest/guide/lifecycle-hooks.html
- https://www.polymer-project.org/1.0/docs/devguide/registering-elements#lifecycle-callbacks
- https://facebook.github.io/react/docs/react-component.html#the-component-lifecycle
- https://developers.google.com/web/fundamentals/getting-started/primers/customelements#reactions

**Comment by mbest** [comment]
Possible list of events for each bound node:

* binding start - before applying bindings to a node
* binding complete - after all bindings for a node have been completed
* node removed - after node is cleaned and removed

* descendants binding start - before applying bindings to a node's descendants
* descendants binding complete - after applying binding to a node's descendants
* descendants removed - after node's bound descendants have been cleaned and removed

**Comment by brianmhunt** [comment]
We should also think about whether the resolution of promises passes as arguments - the valueaccessor - should trigger events.  (fhenables in dustjs are my inspiration for this)

**Comment by mbest** [comment]
@brianmhunt Right. There should be a mechanism for bindings to complete asynchronously.

**Comment by cervengoc** [comment]
@mbest What is the status of implementing the "async" binding handler `init`s? 

If I understand correctly, all exported `applyBinding***` methods should then return a promise to have a standardized interface, even if none of the `init`s in the tree returned a promise object.

**A rough implementation proposal**

- Each `applyBinding***` should return a promise, which is resolved when the bindings are finished underneath it (not taking the parent bindings' possible promises into account).
- We could have a stack of arrays of promises in the `bindingContext` which would inherit though the binding tree by reference. Each array of the stack could represent an external call to one of the `applyBinding**` methods.
- Each time an `applyBinding***` method is called, a new array could be added to the stack, and the combined promises should be returned as one promise (aka. `when`).
- Each time an `init` returns a promise, that should be added to each array of the current stack.

I've implemented a very similar logic locally in my version, because I needed this feature for remotely loaded templates. It seems to be working fine.

Please let me know about your opinion, or if you're already working on this, then please ignore my points.

**Comment by cervengoc** [comment]
Another area where being async could be supported is when one implements a custom template source for providing templates asynchronously like for example the external templates plugin did, or by simply using an AMD loader. I know components already support being async, but not everyone would like to use components I think, many people prefer pure templates, but use some kind of async loading mechanism. 

In these cases the template source could return a promise too, and the template binding should take it into account.

**Comment by brianmhunt** [comment]
I've added a base LifeCycle class (in tko.lifecycle) to tko.

Before it gets to ko-Components, I intend to add an `AsyncLifeCycle` base class.
